File: src/main/java/com/github/bjoernpetersen/jmusicbot/playback/FilePlaybackFactory.java
package com.github.bjoernpetersen.jmusicbot.playback;

import java.io.File;
import java.io.IOException;
import javax.annotation.Nonnull;
import javax.sound.sampled.UnsupportedAudioFileException;

/**
 * An extension of the {@link PlaybackFactory} interface which accepts input files.
 */
public interface FilePlaybackFactory extends PlaybackFactory {

  /**
   * <p>Creates a playback object from the given input file.</p>
   *
   * <p>This method can perform blocking IO actions.</p>
   *
   * @param inputFile the input file with audio data
   * @return a Playback object
   * @throws UnsupportedAudioFileException if the format of the input stream is unsupported
   * @throws IOException if any IO error occurs
   */
  @Nonnull
  Playback createPlayback(@Nonnull File inputFile)
      throws UnsupportedAudioFileException, IOException;
}

File: src/main/kotlin/net/bjoernpetersen/musicbot/spi/plugin/predefined/FilePlaybackFactory.kt
package net.bjoernpetersen.musicbot.spi.plugin.predefined

import net.bjoernpetersen.musicbot.spi.plugin.Playback
import net.bjoernpetersen.musicbot.spi.plugin.PlaybackFactory
import java.io.File
import java.io.IOException

/**
 * An extension of the [PlaybackFactory] interface which accepts input files.
 */
interface FilePlaybackFactory : PlaybackFactory {

    /**
     * Creates a playback object from the given input file.
     *
     * @param inputFile the input file with audio data
     * @return a Playback object
     * @throws UnsupportedAudioFileException if the format of the input stream is unsupported
     * @throws IOException if any IO error occurs
     */
    @Throws(UnsupportedAudioFileException::class, IOException::class)
    suspend fun createPlayback(inputFile: File): Playback
}
